doc/developer/design/20231027_refresh_every_mv.md

```diff
@@ -0,0 +1,282 @@
+   - When we have robust [out-of-core in Compute](https://materializeinc.slack.com/archives/C04UK7BNVL7/p1700084826815849) (probably with auto-scaling), we might be able to trade off freshness vs. cost by simply using a much smaller replica than the cluster's total memory needs, but supplying input data at a much larger tick interval than the default 1 sec. For example, we might supply new input data at every 10 mins, and hope that even with a lot of swapping, we'll be able to process the update in 10 mins.
+  - We could suspend replica processes and save their entire state to S3 and a local disk cache, and restore it when a refresh is needed. [There are existing techniques for saving process state, e.g., CRIU](https://faun.pub/kubernetes-checkpointing-a-definitive-guide-33dd1a0310f6).
+
+We are not aiming to provide hard guarantees that we'll refresh a materialized view at the specified time. This is because the solution approaches considered here are based on spinning up replicas near the refresh times, which is not always possible due to AWS capacity issues. Therefore, timely completion of refreshes are only on a best-effort basis, with possibly tunable knobs by power users that control the trade-off of cost vs. chances that refreshes will complete on time.
+
+The below proposal only considers timelines that have totally ordered timestamps, and whose timestamps relate to wall clock time (as opposed to, e.g., transaction sequence ids of some system).
+
+## Solution Proposal
+
+### `REFRESH` options syntax
+
+We propose providing users various `REFRESH` options on materialized views:
+```
+CREATE MATERIALIZED VIEW ... [WITH (
+  [REFRESH [=] {
+    ON COMMIT
+    | EVERY <every_interval> [STARTING AT <start_timestamp>]
+    | ON SCHEDULE '<cron>'
+    | AT {CREATION | <timestamp>}
+  }]+
+)]
+```
+
+For example:
+```SQL
+CREATE MATERIALIZED VIEW mv1 WITH (
+    REFRESH AT CREATION,
+    REFRESH EVERY '1 day' STARTING AT '2023-11-17 03:00'
+) AS SELECT ...
+```
+
+### `REFRESH` options semantics
+
+Some of the `REFRESH` options can be combined and/or supplied multiple times. (If two refreshes would fall on the exact same moment, then only one refresh is performed at that moment.) The meaning of the various refresh options are as follows.
+
+#### Refresh `ON COMMIT`
+
+This is the default when not supplying any options, i.e., our current behavior of refreshing the materialized view when any changes to our inputs are committed. (The term `REFRESH ON COMMIT` is present in various other database systems, so should be familiar to users.) It is an error to specify this together with any of the other `REFRESH` options.
+
+#### Refresh `EVERY <every_interval> [STARTING AT <start_timestamp>]`
```

Comment:
> I presume that `AT <timestamp>` will error if `<timestamp>` is before the current compaction frontier

Actually, I was also planning to accept and ignore `AT <timestamp>` in the past, for the same reason why we want to accept `STARTING AT` in the past. E.g., for the specific case of [redacted customer name], they will have something like `REFRESH AT CREATION, REFRESH AT '2024-01-01'`, and on another mat view `REFRESH AT CREATION, REFRESH AT '2025-01-01'`, and then maybe in 2026 they suddenly do a full DBT redeploy for some reason, and then we don't want to force them to fiddle with the old `REFRESH AT` times, but just want the `REFRESH AT CREATION`s to take effect. How does this sound?

> What about something like `ALIGNED TO`? I think that does a nice job of being descriptive while not implying anything about starting at that timestamp.

Hmm, sounds ok to me! I'll write in the channel to see how people feel about the clarity of `ALIGNED TO`.

Reply:
Btw. there is one completely different approach we could take (inspired by [Flink's windowing syntax](https://nightlies.apache.org/flink/flink-docs-release-1.19/docs/dev/table/sql/queries/window-tvf/#window-offset), which I checked just now after @uce mentioned it): Instead of specifying an absolute point in time, we could specify an offset. The offset would be relative to the default phase, which could be a refresh at the start of the Unix epoch. 

This would work nicely for some common refresh intervals, e.g., `REFRESH EVERY '1 day' OFFSET '3 hours'` would refresh every day at 3am, `REFRESH EVERY '1 hour' OFFSET '10 minutes'` would refresh every hour at 10 minutes after the hour, and so on. For these intervals, you don't even need to know that the default phase is a refresh at the start of the Unix epoch; you just get the behavior that you intuitively expect.

However, it gets a bit confusing for some other intervals:
- It breaks down for crazy intervals that don't evenly divide days/hours/weeks/etc., e.g., for `REFRESH EVERY '91 minutes' OFFSET '10 minutes'` it's non-trivial to figure out when your first refresh will be. (You'd need to do some calculation involving the Unix epoch.) But this doesn't feel like a big problem, because in such a case it's not clear why you even want to specify the phase at all. Maybe you want such a crazy interval if some other system is giving you data every 91 minutes? This doesn't sound like a common case.
- However, I can see one possibly common case where the offset approach, sadly, breaks down: If you type `REFRESH EVERY '1 week' OFFSET '2 days 4 hours'` you might naively expect that it would refresh every week at Wednesday 4am. However, 1970 Jan. 1. was a Thursday (instead of a Monday), so you get refreshes every Saturday 4am. We could fix this specific problem by choosing a default phase that falls on a Monday midnight. But this issue makes this offset approach feel not so satisfying from a conceptual point of view.

Reply:
> How does this sound?

I'm not entirely sold on it. It's easier to move from restrictive to permissive semantics than vice versa. If we start with `AT` rejecting times in the past and users find that frustrating, it's easy to later decide to allow times in the past. But if we start with `AT` allowing times in the past and users find that confusing, it's backwards incompatible to start rejecting times in the past.

We're allowing `AT` to be an expression that references `mz_now()`, right? I feel like, to get the behavior you describe of ignoring times in the past, we could just tell those users to write:

```sql
REFRESH AT CREATION,
REFRESH AT greatest(mz_now(), '2024-01-01),
```

If executed in 2023, that causes two refreshes: one at creation and one at January 1, 2024. If executed in 2024+, `AT CREATION` and `AT max(mz_now(), '2024-01-01)` both refer to the same instant, and only one refresh is performed at creation.

It's a bit more typing, but I really like how explicit it is.

Reply:
> `REFRESH AT greatest(mz_now(), '2024-01-01')`

Good idea! My only slight concern is that when users initially create the MV, maybe they don't yet know that they want that behavior but they very much do. I guess we can alleviate this concern by explaining this carefully in the docs, and also by hinting this in the error msg that rejects the past `REFRESH AT`.

But note that this trick only works for `REFRESH AT`, but not for `STARTING AT`, because the `mz_now()` branch kicking in would alter the phase.